eZ/Publish/Core/REST/Server/Input/Parser/Criterion/LogicalAnd.php

```diff
@@ -30,19 +35,65 @@ class LogicalAnd extends CriterionParser
      */
     public function parse(array $data, ParsingDispatcher $parsingDispatcher)
     {
-        if (!array_key_exists('AND', $data) && !is_array($data['AND'])) {
-            throw new Exceptions\Parser('Invalid <AND> format');
+        if (!array_key_exists(static::TAG_NAME, $data) || !is_array($data[static::TAG_NAME])) {
+            throw new Exceptions\Parser('Invalid <' . static::TAG_NAME . '> format');
         }
 
         $criteria = array();
-        foreach ($data['AND'] as $criterionName => $criterionData) {
-            if (is_array($criterionData) && !array_key_exists(0, $criterionData)) {
-                $criterionName = key($criterionData);
-                $criterionData = current($criterionData);
+
+        $flattenedCriteriaElements = $this->getFlattenedCriteriaData($data[static::TAG_NAME]);
+        foreach ($flattenedCriteriaElements as $criterionElement) {
+            $criteria[] = $this->dispatchCriterion(
+                $criterionElement['type'],
+                $criterionElement['data'],
+                $parsingDispatcher
+            );
+        }
+
+        return new Values\Content\Query\Criterion\LogicalAnd($criteria);
+    }
+
+    /**
+     * @param array $criteriaByType
+     * @return array
+     */
+    protected function getFlattenedCriteriaData(array $criteriaByType): array
+    {
+        $criteria = [];
+        foreach ($criteriaByType as $type => $criterion) {
+            if (is_array($criterion) && $this->isNumericArray($criterion)) {
```

Comment:
Hmm, this is usually called zero-indexed array. But could you elaborate on what happens here, so this code is required?
_(this is just a question, not change request. I need to have full understanding of what happens here and right now I don't follow :sweat_smile: )._

Reply:
Hey. 

Of course. When parsing a view-input query
```
<Query>
  <OR>
    <Criterion1>some-value</Criterion1>
    <Criterion2>some-value</Criterion2>
  </OR>
</Query>
```
EZ will not parse this into a numeric array instead create an associative array with criteria name (Criterion1 & Criterion2) as indexes.
```
[
  "Criterion1" => "some-value",
  "Criterion2" => "some-value"
]
```
This is a problem when there is two criteria of same type, like:
```
<Query>
  <OR>
    <ContentTypeIdentifierCriterion>author</ContentTypeIdentifierCriterion>
    <ContentTypeIdentifierCriterion>book</ContentTypeIdentifierCriterion>
    <SomeOtherCriterion>some-value</SomeOtherCriterion>
  </OR>
</Query>
```
The decoder will transform this into a nested array:
```
[
  "ContentTypeIdentifierCriterion" => [
    0 => "author",
    1 => "book",
  ],
  "SomeOtherCriterion" => "some-value"
]
```
The method transforms this into a flattend criterion=>configuration array:
```
[
  0 => [
     "type" => "ContentTypeIdentifierCriterion",
     "data" => "author"
  ],
  1 => [
     "type" => "ContentTypeIdentifierCriterion",
     "data" => "book"
  ],
  2 => [
     "type" => "SomeOtherCriterion",
     "data" => "some-value"
  ],
]
```
In case you have a better explaining name for the method please tell me and i will change that.

Reply:
> This looks promising :)
> 
> In general things to change:
> 
> 1. This PR should rather target `6.7` branch (LTS) as bug probably exists there also.
> 2. While the tests cases you've implemented are quite valuable, in case of REST, foremost we need some integration/functional tests in `eZ/Bundle/EzPublishRestBundle/Tests/Functional` namespace.
> 
> Other remarks and questions:

I will add a functional test and change the target branch (probably tomorrow).

Reply:
> I will add a functional test and change the target branch (probably tomorrow).

👍 that will alos help understanding this

Reply:
@andrerom @alongosz 

I changed the target version to 6.7 and implemented functional tests for the View POST route and LogicalAnd, LogicalOr and nesting of those.

Unfortunately `continuous-integration/appveyor/` says it has merge conflicts because while reconfiguring the PR it was still going to master. I changed it but don't know how to trigger appveyor again.

[ ] PHP Code Fixer applied
[ ] Functional Tests implemented
[ ] Unit Tests implemented
[ ] Bug fixed
[ ] Target version is now 6.7

Side notice:
I realized that on 6.7 the FieldCriterion Parser is not yet implemented and to curiosity backported the current one to it and it works perfect with all tests passing, though I was not sure if it is a good idea to add this feature on this bugfix branch. If you think that backporting the FieldCriterion is an option i would rather merge this:
https://github.com/RozbehSharahi/ezpublish-kernel/commits/backup/fix/rest-api-logical-operators

So people can also use the FieldCriterion on REST and 6.7.

Reply:
@alongosz WDYT?

Reply:
> I realized that on 6.7 the FieldCriterion Parser is not yet implemented and to curiosity backported the current one to it and it works perfect with all tests passing, though I was not sure if it is a good idea to add this feature on this bugfix branch. 

No, we're not backporting features unless we have to due to technical constraints :)